server/server.go

```diff
@@ -239,12 +242,15 @@ func NewServer(cfg *config.Config, driver IDriver) (*Server, error) {
 					go s.forwardUnixSocketToTCP()
 				}
 			}
+			if runInGoTest && s.cfg.Port == 0 {
+				s.cfg.Port = uint(s.listener.Addr().(*net.TCPAddr).Port)
+			}
 		}
 	} else if cfg.Socket != "" {
 		if s.listener, err = net.Listen("unix", cfg.Socket); err == nil {
 			logutil.BgLogger().Info("server is running MySQL protocol", zap.String("socket", cfg.Socket))
 		}
-	} else {
+	} else if !(s.cfg.Port == 0 && runInGoTest) {
```

Comment:
```suggestion
	} else  {
```
I think it should raise an error whenever `s.listener` is not bind.

Reply:
Or `s.listener` would be nil after this function returns.